### ISSUE TITLE
Clarification on extracting learned margins from Phase One in CLOC training

### ISSUE THREAD

**Issue by teinhonglo** [dpitawela/CLOC#1]
Hi, thank you for sharing this great work!

I have a question regarding the two-phase training strategy described in the README and implemented in `OrdinalContrastiveLoss_mm` and `OrdinalContrastiveLoss_sm`. Specifically, I am unclear on **how to extract and reuse the margin values learned in Phase One** for use in Phase Two.

From the documentation:

> Phase One: model parameters and the parameters of the OrdinalContrastiveLoss (i.e. the margins) are optimized jointly.
> Phase Two: only the model parameters are optimized, and the learned margins from Phase One are reused via `learnable_map`.

I have already completed training using `OrdinalContrastiveLoss_mm` and `OrdinalContrastiveLoss_sm`, but I am not sure how to **retrieve the learned margin values** after Phase One. Should I:

1. Access `loss.learnables.data` (for `OrdinalContrastiveLoss_mm`) or `loss.distance.data` (for `OrdinalContrastiveLoss_sm`) directly after Phase One?
2. Manually convert them back via `softplus`?
3. Construct the `learnable_map` for Phase Two like:

   ```python
   learnable_map = [('fixed', learned_margin_1), ('fixed', learned_margin_2), ...]
   ```

If possible, could you provide a short code snippet or guidance on the correct way to perform this transition?

Thanks again for your time and this excellent implementation!

Best regards,
Tien-Hong

**Comment by dpitawela** [comment]
Hi Tien-Hong, Thank you for the question.

**Option 1: Straight Forward Method** 

If you pass a tensorboard ```summaryWriter``` instance to loss object when initialising it, and pass ```step``` to the forward method. It should generate graphs where you can see the margin values with tensorboard.

**Option 2:**

```F.softplus(loss.learnables.data)``` when using ```OrdinalContrastiveLoss_mm```
```F.softplus(loss.distance.data)``` when using ```OrdinalContrastiveLoss_sm```

But have to do this before the program terminates.

------
The accessed values using either of the options above can be used to create the learnable map for phase two training.

I'll update the repository with the description above. Thanks again for the question.

**Comment by teinhonglo** [comment]
Hi, thank you for your helpful response!

Just a quick follow-up to confirm that my current implementation aligns with the intended usage.

In my implementation of `OrdinalContrastiveLoss_mm`, I slightly modified the `__getGraphLayout` function as follows to visualize the learned margins:

```python
def __getGraphLayout(self):
    layout = {
        "Class Distances": {
            "margin": ["Multiline", [f'C{i}-C{i+1} {dist_val}' for i, dist_val in enumerate(self.distances_ori)]],
        }
    }
    return layout
```

Here, `self.distances_ori` is derived from `F.softplus(self.learnables).data` before training ends. I use this to display margin values in TensorBoard.

Could you kindly confirm if this is the correct or recommended way to visualize the margins in Phase One?

Thanks again for your guidance!

Best regards,
Tien-Hong

**Comment by dpitawela** [comment]
Hi Tien-Hong,

Thanks for the follow-up.
```__getGraphLayout()``` just creates the layout for the graph and only runs once. The logging is done in ```__logDistances()``` and called in every iteration.

If you want to change how the graph look like, update: ```__getGraphLayout()```
If you want to change what to log, update: ```__logDistances()```.

Additional: In my implementation, the value on the tensorboard graph can be directly used to create the ```learnable_map ``` for the phase 2. Hope this helps.

I'm sorry, without the whole implementation, I cannot confirm if yours is doing the same.

**Comment by teinhonglo** [comment]
Hi, thank you again for your helpful response regarding Phase One margin extraction.

I would like to follow up with a question about logging. In my current implementation of `OrdinalContrastiveLoss_mm`, I log the learned margins using Python’s `logging` module instead of `SummaryWriter`. Specifically, I use the following implementation:

```python
def __logDistances(self, step, distances):
    if self.writer != None and step != None:
        distance_info = [f"C{i}-C{i+1}={dist:.4f}" for i, dist in enumerate(distances)]
        self.writer.info(f"Step {step}: " + ", ".join(distance_info))
```

Here, `distances` is obtained from `F.softplus(self.learnables)` during Phase One training. These values are written into a text-based log file for inspection.

Could you kindly confirm whether this approach is acceptable for logging and reusing the margin values in Phase Two? Or would you recommend using `SummaryWriter.add_scalar()` to visualize the margin evolution in TensorBoard instead?

For reference, here is the relevant portion of my implementation:

```python
class OrdinalContrastiveLoss_mm(nn.Module):
    def __init__(self, n_classes, learnable_map=None, writer=None):
        super().__init__()
        self.n_distances = n_classes - 1

        if writer is not None:
            logging.basicConfig(
                filename=writer,
                level=logging.INFO,
                format='%(asctime)s - %(levelname)s - %(message)s'
            )
            self.writer = logging.getLogger("DistanceLogger")

        self.__createLearnableTensor(learnable_map)

    def __createLearnableTensor(self, learnable_map):
        if learnable_map is None:
            learnable_map = [['learnable', None] for _ in range(self.n_distances)]

        self.distances_ori = torch.zeros(self.n_distances, dtype=torch.float64)
        learnable_indices = []

        for i, (isFixed, value) in enumerate(learnable_map):
            if isFixed == 'learnable':
                self.distances_ori[i] = self.__inverse_softplus(
                    0.5 + torch.rand(1) * 0.5 if value is None else torch.tensor([value])
                )
                learnable_indices.append(i)
            elif isFixed == 'fixed':
                self.distances_ori[i] = self.__inverse_softplus(torch.tensor([value]))

        if learnable_indices:
            learnable_indices = torch.tensor(learnable_indices)
            self.learnables = nn.Parameter(self.distances_ori[learnable_indices])
            self.mask_learnables = torch.zeros_like(self.distances_ori, dtype=torch.bool)
            self.mask_learnables[learnable_indices] = True

    def __contrastiveLoss(self, prediction, target, step=None, device=None):
        self.distances = self.distances_ori.clone().to(device)
        if hasattr(self, 'learnables'):
            self.learnables = self.learnables.to(device)
        if hasattr(self, 'mask_learnables'):
            self.mask_learnables = self.mask_learnables.to(device)
            self.distances[self.mask_learnables] = self.learnables

        n_samples = prediction.size(0)
        cos_sim = F.cosine_similarity(prediction.unsqueeze(0), prediction.unsqueeze(1), dim=2)
        label_diff = torch.abs(target.unsqueeze(0) - target.unsqueeze(1)).float()
        positives = label_diff <= 0
        negatives = ~positives
        positives[torch.eye(n_samples).bool()] = False

        pos_cossim, neg_cossim = cos_sim.clone(), cos_sim.clone()
        pos_cossim[~positives] = torch.inf
        neg_cossim[~negatives] = -torch.inf

        pos_distances = F.softplus(self.distances)
        class_positions = torch.cumsum(torch.cat([torch.tensor([0.0]).to(device), pos_distances]), dim=0)
        distance_matrix = torch.abs(class_positions.unsqueeze(0) - class_positions.unsqueeze(1))

        self.__logDistances(step, pos_distances)

        label_indices = target.unsqueeze(0).repeat(n_samples, 1).to(device)
        margins = distance_matrix[label_indices, label_indices.t()]
        margins[~negatives] = 0

        mean_n_pair_loss = torch.tensor([]).to(device)
        loss_masks_2 = torch.tensor([]).to(device)

        for pos_col in pos_cossim.T:
            n_pair_loss = (-pos_col + neg_cossim.T + margins.T).T
            loss_mask1 = ~torch.isinf(n_pair_loss)
            loss_mask2 = loss_mask1.sum(dim=1)
            n_pair_loss = F.relu(n_pair_loss)
            n_pair_loss = (n_pair_loss * loss_mask1).sum(dim=1) / loss_mask2.clamp(min=1)
            loss_masks_2 = torch.cat((loss_masks_2, loss_mask2.reshape(1, -1)), dim=0)
            mean_n_pair_loss = torch.cat((mean_n_pair_loss, n_pair_loss.reshape(1, -1)), dim=0)

        mean_n_pair_loss = (mean_n_pair_loss * loss_masks_2.bool()).sum(dim=0) / loss_masks_2.sum(dim=0).clamp(min=1)
        return mean_n_pair_loss.mean()

    def __logDistances(self, step, distances):
        if self.writer is not None and step is not None:
            distance_info = [f"C{i}-C{i+1}={dist:.4f}" for i, dist in enumerate(distances)]
            self.writer.info(f"Step {step}: " + ", ".join(distance_info))

    def __inverse_softplus(self, t):
        return torch.where(t > 20, t, torch.log(torch.exp(t) - 1))

    def forward(self, prediction, target, step=None, device=None):
        return self.__contrastiveLoss(prediction, target, step, device)
```

=========== Updated ===================

One more clarification if I may:

In my current use case, the input to `OrdinalContrastiveLoss_mm` is:

* `prediction`: a tensor of shape **(B, C)**, which corresponds to **logits for C classes**.
* `target`: a tensor of shape **(B, )**, where each value is an integer label in the range **\[0, C-1]**.

This differs from the typical use case where `prediction` might be an embedding vector rather than class logits.

Would you recommend transforming the logits (e.g., via normalization or projection) before applying the contrastive loss? Or is this usage supported as-is?

I’m asking this because `F.cosine_similarity(prediction.unsqueeze(0), prediction.unsqueeze(1), dim=2)` is applied directly, and I want to confirm whether any normalization (e.g., `F.normalize(prediction, dim=1)`) is expected beforehand.

My goal is to apply the ordinal contrastive loss directly after a classifier head without requiring an intermediate embedding layer. I’d like to confirm whether this aligns with the original intent of the loss function design.

Thanks again for your time and for maintaining such a great resource!


Best regards,
Tien-Hong
